### PR TITLE
filter webhooks by selected namespace

### PR DESCRIPTION
**Pull request by carlos-logro**
issue: #61 

# Changes

This PR depends on a PR from the dashboard (https://github.com/tektoncd/dashboard/pull/354) where the selected namespace is passed as a prop to WebhookApp container which then passes the namespace to the WebhookTable. In the table, the namespace helps filter out the webhooks while also using to see if the namespace column needs to be shown in the table. Furthermore, if a webhook is not created under a certain namespace or all namespaces, a message will be shown to the user. Lastly, all test were updated & the test cases added were in the WebhookTable.test file.

### CODE REVIEW

**Comment by carlos-logro** [comment]
/hold cancel /cc @dibbles @AlanGreene 

added package-lock file because the unit tests were not passing without it although i did add the package originally.

**Comment by carlos-logro** [comment]
added package-lock file because the unit tests were not passing without it although i did add the package originally.

File: webhooks-extension/src/components/WebhookDisplayTable/WebhookDisplayTable.js [hunk 1]
```diff
@@ -171,37 +172,50 @@ export class WebhookDisplayTable extends Component {
           <Redirect to={this.props.match.url + "/create"} />
         )
       } else {
+        const { namespace } = this.props;
         // There are webhooks so display table
         const headers = [
           {
-            key: 'name',
-            header: 'Name',
+            key: "name",
```
Comment (AlanGreene) [inline webhooks-extension/src/components/WebhookDisplayTable/WebhookDisplayTable.js]: Why the change to double quotes here? Consistent quote style should be enforced by the linter.

File: webhooks-extension/src/components/WebhookDisplayTable/WebhookDisplayTable.js [hunk 2]
```diff
@@ -171,37 +172,50 @@ export class WebhookDisplayTable extends Component {
...
+        let initialRows = [];
         // Populate the data for the rows array from the data from the webhooks get request made on page load
-        this.state.webhooks.map(function (webhook, keyIndex) {
-          initialRows[keyIndex] = {
-            id: webhook['name']+"|"+webhook['namespace'],
-            name: webhook['name'],
-            repository: webhook['gitrepositoryurl'],
-            pipeline: webhook['pipeline'],
-            namespace: webhook['namespace'],
+        this.state.webhooks.map(function(webhook) {
```
Comment (AlanGreene) [inline webhooks-extension/src/components/WebhookDisplayTable/WebhookDisplayTable.js]: `map` should return an array, if you're just using it to iterate use `forEach` instead

File: webhooks-extension/src/components/WebhookDisplayTable/WebhookDisplayTable.js [hunk 3]
```diff
@@ -171,37 +172,50 @@ export class WebhookDisplayTable extends Component {
...
+              id: webhook["name"] + "|" + webhook["namespace"],
+              name: webhook["name"],
+              repository: webhook["gitrepositoryurl"],
+              pipeline: webhook["pipeline"]
+            });
           }
-        })
+
+          if (namespace === ALL_NAMESPACES) {
+            initialRows[initialRows.length - 1].namespace =
```
Comment (AlanGreene) [inline webhooks-extension/src/components/WebhookDisplayTable/WebhookDisplayTable.js]: it would be clearer to set this before pushing the row onto the array

File: webhooks-extension/src/components/WebhookDisplayTable/WebhookDisplayTable.js (hunk 2 above)
Comment (AlanGreene) [inline webhooks-extension/src/components/WebhookDisplayTable/WebhookDisplayTable.js]: could also deconstruct the `webhook` param here since you only ever need individual fields, avoiding all the `webhook[foo]`

```js
({ gitrepositoryurl, name, namespace, pipeline }) => {

}
```

File: webhooks-extension/src/components/WebhookDisplayTable/WebhookDisplayTable.js [hunk 4]
```diff
@@ -280,6 +294,23 @@ export class WebhookDisplayTable extends Component {
               />
             </div>
 
+            {initialRows.length === 0 && namespace === ALL_NAMESPACES && (
```
Comment (AlanGreene) [inline webhooks-extension/src/components/WebhookDisplayTable/WebhookDisplayTable.js]: Refactor to avoid duplicating the containers, it's only the strings that differ in these 2 cases. This will help to ensure consistency after any future updates as there's only a single instance of the container that needs to be updated.

```js
{initialRows.length === 0 && (
  <p className=...>
    {namespace === ALL_NAMESPACES ? 'No webhooks created under any ...' : 'No webhooks created under namespace ...'}
  </p>
)}
```

File: webhooks-extension/src/components/WebhookDisplayTable/tests/WebhookTable.test.js [hunk 5]
```diff
@@ -24,22 +24,29 @@ describe('without webhooks', () => {
...
     jest.spyOn(API, 'getWebhooks').mockImplementation(() => Promise.resolve([noWebhooks]));
-    const { queryByText, queryByTestId } = renderWithRouter(<WebhookDisplayTable match={{}} />);
+    const { queryByText, queryByTestId } = renderWithRouter(<WebhookDisplayTable match={{}} namespace="*"/>);
     expect(queryByTestId('webhook-notification')).toBeNull();
     expect(queryByText(/Loading/i)).toBeTruthy();
   });
-}); 
+
+  it('should display string suggesting to create webhooks when no namespace has webhooks', () => {
+    const noWebhooks = [ ];
```
Comment (AlanGreene) [inline webhooks-extension/src/components/WebhookDisplayTable/tests/WebhookTable.test.js]: there's no reference to this 'string suggesting to create webhooks' in the test, this looks identical to the test above

File: webhooks-extension/src/WebhookApp.js [hunk 6]
```diff
@@ -25,15 +26,26 @@ class WebhooksApp extends Component {
...
 }
 
-export default withRouter(WebhooksApp)
+function mapStateToProps(state, props) {
+  return {
+    namespace: props.selectors.getSelectedNamespace(state)
+  };
+}
+
+const mapDispatchToProps = {};
```
Comment (AlanGreene) [inline webhooks-extension/src/WebhookApp.js]: you can omit this from the call to `connect` since it's not being used

File: webhooks-extension/src/components/WebhookDisplayTable/tests/WebhookTable.test.js (hunk 5 above)
Comment (carlos-logro) [reply webhooks-extension/src/components/WebhookDisplayTable/tests/WebhookTable.test.js]: oops must of not pushed that part

Comment (carlos-logro) [reply webhooks-extension/src/components/WebhookDisplayTable/tests/WebhookTable.test.js]: removed after noticing that when there are no webhooks under any namespace the page redirects to the creating modal.

**Comment by carlos-logro** [comment]
@AlanGreene updates done here.
